go-controller/pkg/ovn/controller/apbroute/network_client.go

```diff
@@ -126,67 +126,37 @@ func (nb *northBoundClient) delAllLegacyHybridRoutePolicies() error {
 // If a set of gateways is given, only routes for that gateway are deleted. If no gateways
 // are given, all routes for the namespace are deleted.
 func (nb *northBoundClient) deleteGatewayIPs(podNsName ktypes.NamespacedName, toBeDeletedGWIPs, _ sets.Set[string]) error {
-	for _, routeInfo := range nb.getRouteInfosForPod(podNsName) {
-		// if we encounter error while deleting routes for one pod; we return and don't try subsequent pods
-		if err := nb.deletePodGWRoutes(routeInfo, toBeDeletedGWIPs, podNsName); err != nil {
-			return err
-		}
-	}
-	return nil
-}
-
-// deletePodGWRoutes removes known exgw routes for a pod via routeInfo for a list of given GW IPs
-func (nb *northBoundClient) deletePodGWRoutes(routeInfo *RouteInfo, toBeDeletedGWIPs sets.Set[string], podNsName ktypes.NamespacedName) error {
-	routeInfo.Lock()
-	defer routeInfo.Unlock()
-	if nb.externalGatewayRouteInfo.GetRouteInfoDeletedStatus(podNsName) {
-		return nil
-	}
-	pod, err := nb.podLister.Pods(routeInfo.PodName.Namespace).Get(routeInfo.PodName.Name)
-	var deletedPod bool
-	if err != nil && apierrors.IsNotFound(err) {
-		// Mark this routeInfo as deleted
-		deletedPod = true
-	}
-	if err == nil {
-		local, err := nb.isPodInLocalZone(pod)
-		if err != nil {
-			return err
-		}
-		if !local {
-			klog.V(4).Infof("APB will not delete exgw routes for pod %s not in the local zone %s", routeInfo.PodName, nb.zone)
-			return nil
+	return nb.externalGatewayRouteInfo.Delete(podNsName, func(routeInfo *RouteInfo) error {
```

Comment:
yeah that is the whole point of `externalGatewayRouteInfo`: it is a thread-safe cache of routeInfos for different pods. So when you call `.Delete` it will get route info from the cache, and pass if to the function we define here as an argument with proper locking. Then if the function returns nil as an error route info will be deleted.

Reply:
Ah I get it. Thanks for the clarification :smile: